enhancements/update/upgrade-without-registry.md

```diff
@@ -0,0 +1,891 @@
+of each OCP component, and prevent merging if it detects that the offending pull
+policy is used. We should consider adding admission in CI for this.
+
+### Graduation Criteria
+
+The feature will ideally be introduced as `Dev Preview` in OpenShift 4.X,
+moved to `Tech Preview` in 4.X+1 and declared `GA` in 4.X+2.
+
+#### Dev Preview -> Tech Preview
+
+- Ability to upgrade a single-node clusters using the `imageSource.sourceType:
+Registry` mode (no bundle, just pinning and pre-loading of the images).
+
+- Availability of the tests that verify the upgrade if single-node clusters.
+
+- Availability of the tests that verify that no OCP component uses the `Always`
+pull policy.
+
+- Obtain positive feedback from at least one customer.
+
+#### Tech Preview -> GA
+
+- Ability to manually detect update bundles.
+
+- Ability to upgrade single-node and multi-node clusters using the
+`imageSource.sourceType: Bundle` mode (with a bundle and pinning and
+pre-loading the images).
+
+- Ability to create bundles using the `oc adm upgrade create bundle` command.
+
+- Availability of the tests that verify the upgrade in single-node and
+multi-node clusters.
+
+- User facing documentation created in
+[https://github.com/openshift/openshift-docs](openshift-docs).
+
+#### After GA
+
+- Ability to automatically detect update bundles.
+
+#### Removing a deprecated feature
+
+Not applicable, no feature will be removed.
+
+### Upgrade / Downgrade Strategy
+
+There are no additional considerations for upgrade or downgrade. The same
+considerations that apply to the cluster version operator in general will also
+apply in this case.
+
+### Version Skew Strategy
+
+This feature will only be usable once the cluster version operator and the
+machine config operator have been upgraded to support it. That upgrade will have
+to be done by other means.
+
+For subsequent upgrades we will ensure that the cluster version operator can
+work with both the old and the new version of the machine config operator.
+
+### Operational Aspects of API Extensions
+
+Not applicable, there are no API extensions.
+
+#### Failure Modes
+
+#### Support Procedures
+
+## Implementation History
+
+There is an initial prototype exploring some of the implementation details
+described here in this [https://github.com/jhernand/upgrade-tool](repository).
+
+## Alternatives
+
+The alternative to this is to make a registry server available, either outside
+or inside the cluster.
+
+An external registry server is the currently supported solution for upgrades in
+disconnected environments. It doesn't require any change, but it is not
+feasible in most of the target environments due to the resource limitations
+described in the [motivation](#motivation) section of this document.
+
+An internal registry server, running in the cluster itself, is a feasible
+alternative for clusters with multiple nodes. The registry server supported by
```

Comment:
maybe.  the orchestration of this as the registry moves between nodes during reboots, and how it pulls its own image, seem like questions that would need to be answered for a self-hosted registry solution.

Reply:
In a multi-node scenario that internal registry is a highly available service. That guarantees that it will be able to pull its own image from another node. We tested that with a regular Quay instance, and works correctly.